### PR TITLE
Update to the latest jenkins plugin version 4.8

### PR DESCRIPTION
**Pull request by fr123k**
* update the pom definition to the jenkins plugin version 4.1
* update the plugin specific code to support plugin version 4.1 and
  be fully compatible with the casc plugin (configuration as code)

The already defined tests are all passing at least most of the time. I also
tested the compatibility of the configuration with the version 3.6.0. But i can't
guaranty full compatibility at this point.

I would also like to take over the maintainer role of this plugin and will follow
the take over guide in the next days.

I just came across those plugin working on my own one that needed a way to
prioritize jobs. Instead of implementing this and i spend the time to update this
plugin to the latest jenkins version in order to use the jenkins casc plugin (configuration as code)

Thanks Brad Larson and Magnus Sandberg for providing amnd mainting this plugin.

### CODE REVIEW

**Comment by JungStar** [comment]
We are very interested to get this plugin working with Configuration as Code. Any chance this will get implemented in the near future?

**Comment by FnTm** [comment]
@fr123k thank you for taking the time to do this! I'm also very interested in using priority sorter with JCasC. Are there any news on that front?

**Comment by pussinboots** [comment]
Upcoming week i will update the PR and hope for green build light.

File: pom.xml [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
-		<groupId>org.jenkins-ci.plugins</groupId>
-		<artifactId>plugin</artifactId>
-		<version>1.580</version>
-	</parent>
-	<artifactId>PrioritySorter</artifactId>
-	<version>3.6.1-SNAPSHOT</version>
+    <parent>
+        <groupId>org.jenkins-ci.plugins</groupId>
+        <artifactId>plugin</artifactId>
+        <version>4.1</version>
```
Comment (oleg-nenashev) [inline pom.xml:6]: Now there is 4.8 with a few more updates

**Comment by oleg-nenashev** [comment]
FTR Ownership transfer request: https://groups.google.com/g/jenkinsci-dev/c/7s2xdiHg0QA

File: pom.xml [hunk 2]
```diff
@@ -33,77 +49,85 @@
...
 			Larson</name> <email>bklarson@gmail.com</email> <roles> <role>developer</role> 
 			<role>retired maintainer</role> </roles> <timezone>-6</timezone> </developer> -->
 	</developers>
 
+	<dependencyManagement>
+		<dependencies>
+			<dependency>
+				<groupId>io.jenkins.tools.bom</groupId>
+				<artifactId>bom-2.235.x</artifactId>
+				<version>11</version>
```
Comment (oleg-nenashev) [inline pom.xml:63]: New release with some updates: https://github.com/jenkinsci/bom/releases/tag/bom-12

```suggestion
				<version>12</version>
```

File: pom.xml [hunk 3]
```diff
@@ -25,6 +31,16 @@
...
+			<id>emsa23</id>
+			<name>Frank Ittermann</name>
+			<email>frank.ittermann@data4life.care</email>
+			<roles>
+				<role>developer</role>
+				<role>maintainer</role>
+			</roles>
+			<timezone>CET</timezone>
+		</developer>
+		<!--  Removed not to show on Wiki <developer>
```
Comment (oleg-nenashev) [inline pom.xml:43]: FTR https://issues.jenkins-ci.org/browse/WEBSITE-658 should solve it by removing dependency on Maven metadata.

**Review by oleg-nenashev — COMMENTED** [review]
Would be also great to clean up the commented code. We have commit history if we ever need to retrieve it.
Otherwise it is nice to have a comment explaining why it is commented

File: src/main/java/jenkins/advancedqueue/JobGroup.java [hunk 4]
```diff
@@ -100,9 +123,40 @@ public void setPriorityStrategy(PriorityStrategy priorityStrategy) {
...
-	private JobGroup() {
+	@DataBoundConstructor
+	public JobGroup(int id, String description, int priority, JobInclusionStrategy jobGroupStrategy, boolean runExclusive, boolean usePriorityStrategies, List<? extends PriorityStrategy> priorityStrategies) {
+		this.id = id;
+		this.description = description;
+		this.priority = priority;
+		this.jobGroupStrategy = jobGroupStrategy;
+		this.runExclusive = runExclusive;
+		this.usePriorityStrategies = usePriorityStrategies;
+		/*
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/advancedqueue/JobGroup.java:137]: Copy-pasted from `newInstance()`? Please either cleanup the comment or add a TODO describing why t is commented out. Also, the entire `newIInstance()` might not longer be needed when you have a constructor

File: src/main/java/jenkins/advancedqueue/JobGroup.java [hunk 5]
```diff
@@ -140,7 +196,7 @@ public int getPriority() {
 	 * @deprecated Used in 2.x now replaced with dynamic {@link JobGroup#jobGroupStrategy}, will return the view
 	 */
 	@Deprecated
-        @CheckForNull
+    @CheckForNull
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/advancedqueue/JobGroup.java:199]: Tabs vs. spaces in this and other files. Maybe it makes sense to just add a bulk refactoring in a follow-up pull request

File: src/main/java/jenkins/advancedqueue/JobGroup.java [hunk 6]
```diff
@@ -100,9 +123,40 @@ public void setPriorityStrategy(PriorityStrategy priorityStrategy) {
 	@Deprecated
 	private String jobPattern = ".*";
 	private boolean usePriorityStrategies;
-	private List<JobGroup.PriorityStrategyHolder> priorityStrategies = new ArrayList<JobGroup.PriorityStrategyHolder>();
+	// private List<JobGroup.PriorityStrategyHolder> priorityStrategies = new ArrayList<JobGroup.PriorityStrategyHolder>();
+	private List<PriorityStrategy> priorityStrategies = new ArrayList<PriorityStrategy>();
 
-	private JobGroup() {
+	@DataBoundConstructor
+	public JobGroup(int id, String description, int priority, JobInclusionStrategy jobGroupStrategy, boolean runExclusive, boolean usePriorityStrategies, List<? extends PriorityStrategy> priorityStrategies) {
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/advancedqueue/JobGroup.java:130]: It does not look like you use this constructor anywhere except data binding. For data binding there is no need to repeat fields in the constrictor if you have `@DataBoundSetter`s defined

File: src/main/java/jenkins/advancedqueue/JobGroup.java [hunk 7]
```diff
@@ -181,16 +240,29 @@ public boolean isUsePriorityStrategies() {
...
+	}
+
+	public List<PriorityStrategy> convertPriorityStrategyHolder(List<PriorityStrategyHolder> priorityStrategyHolders) {
+		List<PriorityStrategy> B = priorityStrategyHolders.stream()
+        .map(holder -> holder.priorityStrategy)
+        .collect(Collectors.toList());
+		return B;
 	}
 
 	/**
```
Comment (oleg-nenashev) [inline src/main/java/jenkins/advancedqueue/JobGroup.java:268]: Most likely, `newInstance()` is no longer required,unless there are internal dependencies on this deprecated databinding mechanism

File: LICENSE.txt [hunk 8]
```diff
@@ -15,9 +16,8 @@ all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
```
Comment (oleg-nenashev) [inline LICENSE.txt:19]: Please do not modify the standard license text. It may trigger alarms in license scanners on the user side

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
I still need to go though the entire code. The changes look good overall, but ithe code does not retain binary compatibility with the previous version. I am also not sure about data compatibility.

What I recommend to do:

* Change the target version to 3.0-SNAPSHOT. It does not make much sense to try retainining binary compatibility here (famous last word? might make sense to check plugins depending on this one).
* Add tests which would load the old configuration file (`@LoadData` annotation in Jenkins Test Harness) and verify that they are properly loaded in the new version. If not, an additiona; migration code or documentation will be needed.

File: src/main/java/jenkins/advancedqueue/JobGroup.java (hunk 5 above)
Comment (pussinboots) [reply src/main/java/jenkins/advancedqueue/JobGroup.java:199]: In the single commit https://github.com/jenkinsci/priority-sorter-plugin/pull/50/commits/89a142e35e8295cc4f2d10f9594f40bb615ae125
all tabs are replaced with whitespaces.

**Comment by pussinboots** [comment]
> Add tests which would load the old configuration file (@LoadData annotation in Jenkins Test Harness) and verify that they are properly loaded in the new version. If not, an additiona; migration code or documentation will be needed.

Under `src/test/resources/jenkins/advancedqueue/test/` there a couple of old configuration files that were not changed and still used in all the different tests.

There is only one test that is skipped now MatrixTest.matrix_and_jobs_with_no_configuration its shaky.

At the moment there are no test for JCasC configuration.

File: src/main/java/jenkins/advancedqueue/JobGroup.java (hunk 4 above)
Comment (pussinboots) [reply src/main/java/jenkins/advancedqueue/JobGroup.java]: The `newInstance` method is used by the `PriorityConfiguration.doPriorityConfigSubmit` to serialize the changes done on the /advanced-build-queue configuration page.

At the moment i am missing knowledge on the how this jelly DataBinding works. If this setup is fixed then i guess this method is obsolete.

Comment (pussinboots) [reply src/main/java/jenkins/advancedqueue/JobGroup.java]: For now i removed the not used full args constructor of the JobGroup class.

File: src/main/java/jenkins/advancedqueue/JobGroup.java (hunk 7 above)
Comment (pussinboots) [reply src/main/java/jenkins/advancedqueue/JobGroup.java]: Will keep newInstance and do a Refactoring to the new DataBound mechanism later.
Can you point me to a good documentation or example code of it.

**Comment by pussinboots** [comment]
@JungStar @FnTm

What you can do is to fork this branch and build the plugin your own and install it in jenkins
manually. An example of a JCasC configuration ofr can be obtained from https://github.com/fr123k/jenkins-aws-pulumi/blob/8d5d55ac89db8685c63bc365bd2b39d424594a32/jenkins/config/casc-config/agents.yaml#L39.

On docker hub there is also a image provided that has the new version of this plugin already installed. For this checkout the following github repository https://github.com/fr123k/jocker (jocker jenkins in docker).

**Comment by oleg-nenashev** [comment]
losing in favor of #51 . Thanks anyway @fr123k !

**Pull request closed without merging**